ddtrace/tracer/textmap.go

```diff
@@ -487,3 +488,125 @@ func (*propagatorB3) extractTextMap(reader TextMapReader) (ddtrace.SpanContext,
 }
+
+const (
+	traceparentHeader = "traceparent"
+	tracestateHeader  = "tracestate"
+)
+
+// propagatorW3c implements Propagator and injects/extracts span contexts
+// using W3C tracecontext/traceparent headers. Only TextMap carriers are supported.
+type propagatorW3c struct{}
+
+func (p *propagatorW3c) Inject(spanCtx ddtrace.SpanContext, carrier interface{}) error {
+	switch c := carrier.(type) {
+	case TextMapWriter:
+		return p.injectTextMap(spanCtx, c)
+	default:
+		return ErrInvalidCarrier
+	}
+}
+
+// injectTextMap propagates span context attributes into the writer,
+// in the format of the traceparentHeader and tracestateHeader.
+// traceparentHeader encodes W3C Trace Propagation version, 128-bit traceID,
+// spanID, and a flags field, which supports 8 unique flags.
+// The current specification only supports a single flag called sampled,
+// which is equal to 00000001 when no other flag is present.
+// tracestateHeader is a comma-separated list of list-members with a <key>=<value> format,
+// where each list-member is managed by a vendor or instrumentation library.
+func (*propagatorW3c) injectTextMap(spanCtx ddtrace.SpanContext, writer TextMapWriter) error {
+	ctx, ok := spanCtx.(*spanContext)
+	if !ok || ctx.traceID == 0 || ctx.spanID == 0 {
+		return ErrInvalidSpanContext
+	}
+	flags := ""
+	p, ok := ctx.samplingPriority()
+	if ok && p >= ext.PriorityAutoKeep {
+		flags = "01"
+	} else {
+		flags = "00"
+	}
+
+	writer.Set(traceparentHeader, fmt.Sprintf("00-%032x-%016x-%v", ctx.traceID, ctx.spanID, flags))
+	// if context priority / origin / tags were updated after extraction,
+	// or the tracestateHeader doesn't start with `dd=`
+	// we need to recreate tracestate
+	if ctx.updated ||
+		(ctx.trace != nil && ctx.trace.propagatingTags != nil && !strings.HasPrefix(ctx.trace.propagatingTags[tracestateHeader], "dd=")) {
+		writer.Set(tracestateHeader, composeTracestate(ctx, p, ctx.trace.propagatingTags[tracestateHeader]))
+	} else {
+		writer.Set(tracestateHeader, ctx.trace.propagatingTags[tracestateHeader])
+	}
+	return nil
+}
+
+// composeTracestate creates a tracestateHeader from the spancontext.
+// The Datadog tracing library is only responsible for managing the list member with key dd,
+// which holds the values of the sampling decision(`s:<value>`), origin(`o:<origin>`),
+// and propagated tags prefixed with `t.`(e.g. _dd.p.usr.id:usr_id tag will become `t.usr.id:usr_id`).
+// All tag keys in the list must have all invalid characters replaced with the underscore.
+// Invalid key characters include characters outside the ASCII range 0x20 to 0x7E, space, comma and equal sign.
+// All tag values in the list must have all invalid characters replaced with the underscore.
+// Invalid value characters include characters outside the ASCII range 0x20 to 0x7E, space,
+// comma(reserved for tracestate list-member separator), semi-colon(reserved for separator between entries),
+// and tilde sign(reserved to represent equals sign).
```

Comment:
These comments don't seem to align with the regexps below, or the RFC. My suggestion would be to make the regexp global to the package, comment each of those individually describing what they are checking, and then use them in this function. You can define them globally right above this function. So for example:

```
var (
  // originValueRegexp matches invalid origin value characters, which are characters outside the ASCII range 0x20 to 0x7E, comma (reserved for tracestate list-member separator), semi-colon (reserved for separateor between entries in the dd list-member), and equals (reserved for tracestate list-member key-value separator)
  originValueRegexp = regexp.MustCompile(",|;|=|[^\\x20-\\x7E]+")

  ... the 2 other regexp vars: tagKeyRegexp and tagValRegexp
)
```